opencog/rule-engine/forwardchainer/ForwardChainer.cc

```diff
@@ -144,19 +147,50 @@ void ForwardChainer::do_chain(ForwardChainerCallBack& fcb,
      init_sources = _as.get_outgoing(hsource);
     else
         init_sources.push_back(hsource);
+
     _fcmem.update_potential_sources(init_sources);
 
-    _fcmem.set_source(fcb.choose_next_source(_fcmem)); //set initial source
     auto max_iter = _configReader.get_maximum_iterations();
 
     while (_iteration < max_iter /*OR other termination criteria*/) {
+
         _log->info("Iteration %d", _iteration);
 
-        do_step(fcb);
+        if(rule_weight.empty())
+        {
+            //! Choose next source.
+            _log->info("[ForwardChainer] setting next source");
+
+            _fcmem.set_source(choose_next_source(_fcmem));
+
+            // Choose matching rules whose input matches with the source.
+            vector<Rule*> matched_rules = choose_rules(_fcmem);
```

Comment:
William's idea and the adopted plan is to choose each rule (by tournament/weight first) and attempt to unify with the selected source one-by-one until a match is found rather than iterating through and attempting to unify all of the rules. Please see the spec doc. Does that make sense?